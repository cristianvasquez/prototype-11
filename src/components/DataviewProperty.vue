<script setup>

const props = defineProps({
  property: {
    type: Object,
    required: true
  },
  labels: {
    type: Object,
    required: false
  }
})

</script>

<template>
  <abbr :title="props.labels">{{ props.property }}</abbr>
</template>
