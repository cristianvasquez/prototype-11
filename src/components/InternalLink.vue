<script lang="ts" setup>
import {hoverPreview, isInVault, openOrSwitch,} from 'obsidian-community-lib'
import {inject} from '@vue/runtime-core'
import {getFileTitle} from '../lib/obsidianHelpers'
import {SIDE_VIEW_ID} from '../main'
import {AppContext} from '../types'
import {PropType} from "vue";
import {ItemView} from "obsidian";

const context: AppContext = inject('context')

const props = defineProps({
  linkTo: {
    type: String as PropType<string>,
    required: true
  }
})

async function open(event: MouseEvent) {
  await openOrSwitch(context.app, props.linkTo, event)
}

async function hover(event: MouseEvent) {
  //@ts-Ignore
  const trickObsidianAPI:ItemView = {
    app: context.app,
    getViewType: () => SIDE_VIEW_ID // 'prototype-11-sideview'
  }

  await hoverPreview(event, trickObsidianAPI, props.linkTo)
}

</script>

<template>
        <span
            class="internal-link"
            @click="open"
            @mouseover="hover"
        >
          <a
              :class="isInVault(context.app, linkTo)?'':'is-unresolved'"
              class="internal-link"
          >{{ getFileTitle(props.linkTo) }}</a
          >
        </span>
</template>