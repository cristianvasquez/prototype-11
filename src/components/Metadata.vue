<script lang="ts" setup>
import MetadataValues from './helpers/MetadataValues.vue'
import {createBuilder} from '../lib/rdfBuilder'
import {inject, PropType} from 'vue'
import {dateTimeFormat} from '../consts'
import {RDFModal} from './RDFModal'
import {App} from 'obsidian'
import {Metadata} from "../lib/helpers";

const app: App = inject('app')

const props = defineProps({
  metadata: {
    type: Object as PropType<Metadata>,
    required: true
  }
})

async function popupRDF() {
  const {metadataToRDF} = createBuilder(app)
  // const data = metadataToRDF(props.metadata, dedupFields.value)
  // new RDFModal(app, data.dataset.toString()).open()
  new RDFModal(app, 'Hi there').open()

}


</script>

<template v-if="metadata">

  <div class="attributes">

    <template v-if="metadata.created">
      <div>Created</div>
      <div>{{ metadata.created.toLocaleString(dateTimeFormat) }}</div>
    </template>

    <template v-if="props.metadata.updated">
      <div>Modified</div>
      <div>{{ metadata.updated.toLocaleString(dateTimeFormat) }}</div>
    </template>

    <template v-if="props.metadata.size">
      <div>Size</div>
      <div>{{ metadata.size }}</div>
    </template>

  </div>

  <template v-if="metadata.pairs">
    <h3>Attributes</h3>
    <div class="attributes">
      <template v-for="tuple in metadata.pairs">
        <metadata-values :value="tuple.predicate"/>
        <metadata-values :value="tuple.object"/>
      </template>
    </div>
  </template>

  <template v-if="metadata.triads">
    <h3>Triples</h3>
    <div class="fields">
      <template v-for="tuple in metadata.triads">
        <metadata-values :value="tuple.subject"/>
        <metadata-values :value="tuple.predicate"/>
        <metadata-values :value="tuple.object"/>
      </template>
    </div>

  </template>

  <!--  <h3 v-if="metadata.links && metadata.links.length">Links</h3>-->
  <!--  <template v-for="(value) in metadata.links">-->
  <!--    <dataview-value :value="value"/>-->
  <!--  </template>-->

  <!--  <h3><a @click="popupRDF">Gimme some RDF</a></h3>-->

  <!--  <h3>Tasks</h3>-->
  <!--  <div v-for="(value) in props.metadata.tasks">-->
  <!--    <dataview-value :value="value"/>-->
  <!--    <br/>-->
  <!--  </div>-->
</template>

<style>

.attributes {
  display: grid;
  grid-template-columns: 1fr 5fr;
  padding: 5px;
  gap: 10px;
}

.fields {
  display: grid;
  grid-template-columns: 2fr 2fr 5fr;
  padding: 5px;
  gap: 5px;
}


</style>