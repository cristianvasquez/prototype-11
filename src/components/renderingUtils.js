function splitTextLink (txt, links) {
  const includesLink = (txt) => {
    const match = links.filter((x) => (txt === `[[${x}]]` || txt.indexOf(`[[${x}]]`) > 0))
    return match.length > 0 ? match[0] : false
  }
  const match = includesLink(txt)
  if (match) {
    const pre = txt.substring(0, txt.indexOf(`[[${match}]]`))
    const result = []
    if (pre.length) {
      result.push({
        value: pre,
        type: 'text'
      })
    }
    result.push({
      value: match,
      type: 'link',
    })
    return [true, result]
  }
  return [false]
}

function getSpans (text, links) {

  let result = []
  let current = 0
  for (let i = 0; i < text.length; i++) {
    let chunk = text.substring(current, i)
    const [isMatch, matches] = splitTextLink(chunk, links)
    if (isMatch) {
      current = i
      result = [...result, ...matches]
    }
    if (i === text.length - 1) { // Add the last chunk
      const tail = text.substring(current, text.length)
      const [isMatch, matches] = splitTextLink(tail, links)
      if (isMatch) {
        result = [...result, ...matches]
      } else {
        result.push({
          value: tail,
          type: 'text'
        })
      }
    }
  }
  return result
}

export { getSpans }