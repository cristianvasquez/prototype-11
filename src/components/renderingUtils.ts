function splitTextLink(txt: string, links: Array<string>) {
    const includesLink = (txt: string) => {
        const match = links.filter((x) => (txt === `[[${x}]]` || txt.indexOf(`[[${x}]]`) > 0))
        return match.length > 0 ? match[0] : false
    }
    const match = includesLink(txt)
    if (match) {
        const pre = txt.substring(0, txt.indexOf(`[[${match}]]`))
        const result = []
        if (pre.length) {
            result.push({
                value: pre,
                type: 'text'
            })
        }
        result.push({
            value: match,
            type: 'link',
        })
        return result
    }
    return null
}

function getSpans(text: string, links: Array<string>) {

    let result: any = []
    let current = 0
    for (let i = 0; i < text.length; i++) {
        let chunk = text.substring(current, i)
        const matches = splitTextLink(chunk, links)
        if (matches) {
            current = i
            result = [...result, ...matches]
        }
        if (i === text.length - 1) { // Add the last chunk
            const tail = text.substring(current, text.length)
            const matches = splitTextLink(tail, links)
            if (matches) {
                result = [...result, ...matches]
            } else {
                result.push({
                    value: tail,
                    type: 'text'
                })
            }
        }
    }
    return result
}

export {getSpans}