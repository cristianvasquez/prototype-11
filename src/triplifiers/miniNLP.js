import nlp from 'compromise'
import dates from 'compromise-dates'
import numbers from 'compromise-numbers'
import {LINKS_REGEXP} from './regexp.js'

nlp.extend(dates).extend(numbers)

const matchFirst = (text, patterns) => {

  const doc = nlp(text)
  for (let current of patterns) {
    const values = []
    const groups = doc.match(current).groups()
    for (const [key, value] of Object.entries(groups)) {
      values.push(value.text())
    }
    if (values.length > 0) {
      return values
    }
  }

  return undefined
}

function getInternalLinks (text) {
  const internalLinks = []
  for (const match of text.matchAll(LINKS_REGEXP)) {
    internalLinks.push(match[1])
  }
  return internalLinks
}

function getNLPstuff (text) {
  const doc = nlp(text)
  const dates = doc.dates().json()

  let result = {}
  if (dates.length) {
    result.dates = dates
  }

  // @ts-ignore
  const numbers = doc.numbers().json()
  if (numbers.length) {
    result.numbers = numbers
  }

  const internalLinks = getInternalLinks(text)
  if (internalLinks.length) {
    result.internalLinks = internalLinks
  }
  return result
}

export { getNLPstuff, getInternalLinks, matchFirst }