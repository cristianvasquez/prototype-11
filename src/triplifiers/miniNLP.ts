import nlp from 'compromise'
import dates from 'compromise-dates'
import numbers from 'compromise-numbers'
import {LINKS_REGEXP} from '../lib/regexp'

nlp.extend(dates).extend(numbers)

const matchFirst = (text: string, patterns: Array<string>) => {

    const doc = nlp(text)
    for (let current of patterns) {
        const values = []
        const groups = doc.match(current).groups()
        for (const [key, value] of Object.entries(groups)) {
            values.push(value.text())
        }
        if (values.length > 0) {
            return values
        }
    }

    return undefined
}

function getInternalLinks(text: string) {
    const internalLinks = []
    for (const match of text.matchAll(LINKS_REGEXP)) {
        internalLinks.push(match[1])
    }
    return internalLinks
}

function extract(text: string) {
    const doc = nlp(text)
    // @ts-ignore
    const dates = doc.dates().json()

    let result = {}
    if (dates.length) {
        // @ts-ignore
        result.dates = dates
    }

    // @ts-ignore
    const numbers = doc.numbers().json()
    if (numbers.length) {
        // @ts-ignore
        result.numbers = numbers
    }

    const internalLinks = getInternalLinks(text)
    if (internalLinks.length) {
        // @ts-ignore
        result.internalLinks = internalLinks
    }
    return result
}

export {extract, getInternalLinks, matchFirst}