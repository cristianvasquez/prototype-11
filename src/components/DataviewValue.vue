<script setup>
import InternalLink from './InternalLink.vue'
import {dateTimeFormat} from '../consts'

const props = defineProps({
  value: {
    type: Object,
    required: true
  }
})

function isFile () {
  return props.value.type === 'file'
}

function isDate () {
  return props.value.isLuxonDateTime
}


</script>

<template>
  <div v-if="isFile()">
    <internal-link :linkTo="props.value.path" />
  </div>
  <div v-else-if="isDate()">
    {{ props.value.toLocaleString(dateTimeFormat) }}
  </div>
  <div v-else>{{ props.value }}</div>
</template>