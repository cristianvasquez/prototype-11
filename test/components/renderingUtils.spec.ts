import expect from 'expect'
import {getSpans} from '../../src/components/renderingUtils'
import {getInternalLinks} from '../../src/triplifiers/miniNLP'

const phrases = [
    '[[b]]',
    'a [[b]] c',
    'a [[b]]',
    '[[b]] c',
    '[[a]] [[b]] [[a]]',
    'a [[b]] c [[d]] e [[b]] d',
    'a [[b]] c [[d]] e [[b]] d',
    '[[a | b]]',
]

describe('[renderingUtils]', function () {
    describe('[getSpans]', function () {
        phrases.forEach((current) => {
            it(`"${current}"`, function () {
                const actual = getSpans(current, getInternalLinks(current))
                expect(actual).toMatchSnapshot()
            })
        })
    })
})