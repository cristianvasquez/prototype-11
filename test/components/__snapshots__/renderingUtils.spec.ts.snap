// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[renderingUtils] [getSpans] "[[a | b]]" 1`] = `
Array [
  Object {
    "type": "link",
    "value": "a | b",
  },
]
`;

exports[`[renderingUtils] [getSpans] "[[a]] [[b]] [[a]]" 1`] = `
Array [
  Object {
    "type": "link",
    "value": "a",
  },
  Object {
    "type": "text",
    "value": " ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " ",
  },
  Object {
    "type": "link",
    "value": "a",
  },
]
`;

exports[`[renderingUtils] [getSpans] "[[b]] c" 1`] = `
Array [
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " c",
  },
]
`;

exports[`[renderingUtils] [getSpans] "[[b]]" 1`] = `
Array [
  Object {
    "type": "link",
    "value": "b",
  },
]
`;

exports[`[renderingUtils] [getSpans] "a [[b]] c [[d]] e [[b]] d" 1`] = `
Array [
  Object {
    "type": "text",
    "value": "a ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " c ",
  },
  Object {
    "type": "link",
    "value": "d",
  },
  Object {
    "type": "text",
    "value": " e ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " d",
  },
]
`;

exports[`[renderingUtils] [getSpans] "a [[b]] c [[d]] e [[b]] d" 2`] = `
Array [
  Object {
    "type": "text",
    "value": "a ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " c ",
  },
  Object {
    "type": "link",
    "value": "d",
  },
  Object {
    "type": "text",
    "value": " e ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " d",
  },
]
`;

exports[`[renderingUtils] [getSpans] "a [[b]] c" 1`] = `
Array [
  Object {
    "type": "text",
    "value": "a ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
  Object {
    "type": "text",
    "value": " c",
  },
]
`;

exports[`[renderingUtils] [getSpans] "a [[b]]" 1`] = `
Array [
  Object {
    "type": "text",
    "value": "a ",
  },
  Object {
    "type": "link",
    "value": "b",
  },
]
`;
